README.md
# mpiP 3.5
A light-weight MPI profiler.

## Introduction
mpiP is a light-weight profiling library for MPI applications. Because it only collects statistical information about MPI functions, mpiP generates considerably less overhead and much less data than tracing tools. All the information captured by mpiP is task-local. It only uses communication during report generation, typically at the end of the experiment, to merge results from all of the tasks into one output file.

## Downloading
The current version of mpiP can be accessed at [https://github.com/LLNL/mpiP/releases/latest](https://github.com/LLNL/mpiP/releases/latest).

## New Features & Bug Fixes
Version 3.5 includes several new features, including

- Multi-threaded support
- Additional MPI-IO functions
- Various updates including
  - New configuration options and tests
  - Updated test suite
  - Updated build behavior

Please see the ChangeLog for additional changes.

##  Configuring and Building mpiP
### Dependencies
- MPI installation
- libunwind : for collecting stack traces.  
- binutils : for address to source translation
- glibc backtrace() can also be usef for stack tracing, but source line numbers may be inconsistent.

### Configuration
Several specific configuration flags can be using, as provided by ```./configure -h```.
Standard configure flags, such as CC, can be used for specifying MPI compiler wrapper scripts.

### Build Make Targets
|Target|Effect|
|---|---|
|[default]| Build libmpiP.so|
|all|Build shared library and all tests|
|check|Use dejagnu to run and evaluate tests|

## Using mpiP
Using mpiP is very simple. Because it gathers MPI information through the MPI profiling layer, mpiP is a link time library. That is, you don't have to recompile your application to use mpiP. Note that you might have to recompile to include the '-g' option. This is important if you want mpiP to decode the PC to a source code filename and line number automatically. mpiP will work without -g, but mileage may vary.

### Instrumentation
#### Link Time Instrumentation
Link the mpiP library with an executable. The dependent libraries may need to be specified as well.  If the link command includes the MPI library, order the mpiP library before the MPI library, as in ``` -lmpiP -lmpi```.

#### Run Time Instrumentation
An uninstrumented executable may able to be instrumented at run time by setting the LD_PRELOAD environment variable, as in ```export LD_PRELOAD=[path to mpiP]/libmpiP.so```.  Preloading libmpiP can possibly interfere with the launcher and may need to be specified on the launch command, such as ```srun -n 2 --export=LD_PRELOAD=[path to mpiP]/libmpiP.so [executable]```.

### mpiP Run Time Flags
The behavior of mpiP can be set at run time through the use of the following flags.  Multiple flags can be delimited with spaces or commas.

|Option|	Description|	Default|
|---|---|---|
|-c	| Generate concise version of report, omitting callsite process-specific detail.||
|-d	|Suppress printing of callsite detail sections.||
|-e|Print report data using floating-point format.  |
|-f dir| Record output file in directory \<dir>. |.|
|-g|Enable mpiP debug mode.| disabled|
|-k n |Sets callsite stack traceback depth to <n>.| 1|
|-l|Use less memory to generate the report by using MPI collectives to generate callsite information on a callsite-by-callsite basis.  |
|-n|Do not truncate full pathname of filename in callsites.  |
|-o|Disable profiling at initialization. Application must enable profiling with MPI_Pcontrol().  |
|-p|Point-to-point histogram reporting on message size and communicator used.|
|-r|Generate the report by aggregating data at a single task. |default|
|-s n|Set hash table size to \<n>. |256|
|-t x|Set print threshold for report, where \<x> is the MPI percentage of time for each callsite. |0.0|
|-v|Generates both concise and verbose report output.  |
|-x exe |Specify the full path to the executable.  |
|-y|Collective histogram reporting on message size and communicator used.|
|-z|Suppress printing of the report at MPI_Finalize.  |
	
For example, to set the callsite stack walking depth to 2 and the report print threshold to 10%, you simply need to define the mpiP string in your environment, as in any of the following examples:

```
$ export MPIP="-t 10.0 -k 2" (bash)

$ export MPIP=-t10.0,-k2 (bash)

$ setenv MPIP "-t 10.0 -k 2" (csh)
```

mpiP prints a message at initialization if it successfully finds the MPIP variable.

For more information on mpiP, please see the User Guide in the mpiP distribution.

## License
Copyright (c) 2006, The Regents of the University of California.
Produced at the Lawrence Livermore National Laboratory
Written by Jeffery Vetter and Christopher Chambreau.
UCRL-CODE-223450.
All rights reserved.
 
This file is part of mpiP.  For details, see http://llnl.github.io/mpiP.
 
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:
 
* Redistributions of source code must retain the above copyright
notice, this list of conditions and the disclaimer below.

* Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the disclaimer (as noted below) in
the documentation and/or other materials provided with the
distribution.

* Neither the name of the UC/LLNL nor the names of its contributors
may be used to endorse or promote products derived from this software
without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OF
THE UNIVERSITY OF CALIFORNIA, THE U.S. DEPARTMENT OF ENERGY OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 
Additional BSD Notice
 
1. This notice is required to be provided under our contract with the
U.S. Department of Energy (DOE).  This work was produced at the
University of California, Lawrence Livermore National Laboratory under
Contract No. W-7405-ENG-48 with the DOE.
 
2. Neither the United States Government nor the University of
California nor any of their employees, makes any warranty, express or
implied, or assumes any liability or responsibility for the accuracy,
completeness, or usefulness of any information, apparatus, product, or
process disclosed, or represents that its use would not infringe
privately-owned rights.
 
3.  Also, reference herein to any specific commercial products,
process, or services by trade name, trademark, manufacturer or
otherwise does not necessarily constitute or imply its endorsement,
recommendation, or favoring by the United States Government or the
University of California.  The views and opinions of authors expressed
herein do not necessarily state or reflect those of the United States
Government or the University of California, and shall not be used for
advertising or product endorsement purposes.



ChangeLog
- 2020-10-20 Update deprecated functions. 
- 2020-10-20 Fix ARM typo and add LSE configure test. 
- 2020-10-19 Correct pt2pt histogram reporting check. 
- 2020-10-19 Remove bugreport info from report banner. 
- 2020-10-19 Check for python 
- 2020-10-19 Address compiler warnings for some mt function return types. 
- 2020-10-19 Correct build behavior for testing multi-threaded support with Spectrum MPI. 
- 2020-10-16 Missing function argument entry. 
- 2020-10-16 Merge pull request #29 from roblatham00/more-mpi-io 
- 2020-10-16 Change default make to build shared library.  Target all also makes tests. 
- 2020-10-15 Additional I/O routines 
- 2020-10-16 currBasetype fix 
- 2020-10-16 Testing updates including expanded rma and mpiio. 
- 2020-10-15 Merge pull request #28 from LLNL/mt-test-suite-updates 
- 2020-10-15 Update to test suite to include multi-threaded tests 
- 2020-10-14 IBM jsrun test support 
- 2020-10-14 Followup for renaming wrappers.c to mpiP-wrappers.c 
- 2020-10-14 Merge pull request #27 from LLNL/clarify-stack-trace-behavior 
- 2020-10-14 Include mpiP in wrappers file name for call site reporting 
- 2020-10-14 Clarify stack tracing behavior 
- 2020-10-10 Add IBM jsrun to configure LAUNCH macro. 
- 2020-10-10 Correct autoheader error with HAVE_BFD_GET_SECTION_MACROS. Make define behavior consistent. 
- 2020-10-09 Merge pull request #23 from LLNL/3.4.1-arg-limit-fix 
- 2020-10-09 Minor configuration changes and refactoring. 
- 2020-10-09 Merge with master. 
- 2020-10-09 Merge pull request #26 from milthorpe/master 
- 2020-10-08 Missing PIC configure macro. 
- 2020-10-07 Handle PIC variants. 
- 2020-10-07 Convenience functionality to avoid SMPI libmpiP 
- 2020-10-07 Arch files corrections for gcc and opal code. 
- 2020-10-06 update to work with binutils-2.34 
- 2020-01-10 Added configure option to set max number of copied command line args. 
- 2020-01-10 Fixed report header printing bug with large argument counts. 
- 2020-01-10 Updated GNU config scripts to support modern architectures. 
- 2019-10-31 Merge pull request #22 from LLNL/make_wrapper_cleanup 
- 2019-10-30 fix make-wrappers.py indentation and make python 2 and 3 compatible 
- 2019-08-29 Re-organized sizeof test to avoid CC defaulting to gcc. Re-organized default compilers. 
- 2019-08-29 Added required macro. 
- 2019-08-29 Merge pull request #17 from artpol84/topic/add_tests 
- 2019-06-29 Add Multi-Threading (MT) test suite 
- 2019-08-29 Merge pull request #19 from artpol84/topic/mt+site 
- 2019-07-27 Fix FORTRAN Bindings: properly handle strings 
- 2019-08-29 Merge pull request #13 from artpol84/topic/mt 
- 2019-07-01 Fix the code accounting the application time 
- 2019-06-27 Add support for 32-bit systems 
- 2019-06-27 Fix TLS cleanup code 
- 2019-06-26 Fix MT timings 
- 2019-06-26 Refine statistics reset procedure 
- 2019-06-24 Reduce the number of TLS accesses 
- 2019-06-19 First cut on multi-threading 
- 2019-06-18 Extract callsite code to a separate module 
- 2019-06-18 Move nested call check down the stack 
- 2019-06-18 Introduce additional statistics layer 
- 2019-06-13 First step towards MT support 
- 2019-08-29 Merge pull request #15 from artpol84/fix/testsuite 
- 2019-08-29 Merge pull request #12 from artpol84/topic/pre_mt_refactor 
- 2019-06-12 Refactor global state: remove unneeded component 
- 2019-06-12 Preliminary step for adding multi-threaded support 
- 2019-06-27 Configury change: introduce mpiP-state.[c|h] files 
- 2019-06-27 Get rid of mpiPi.h.in file as unneeded 
- 2019-06-27 Fixes in the configury logic 
- 2019-06-27 Merge pull request #11 from artpol84/fix/indent 
- 2019-06-27 Adding autoconf PACKAGE defines to address current bfd.h requirement. 
- 2019-06-26 Fix mpi.h includes in some of the tests 
- 2019-06-07 Indentation fixes only. No actual changes. 
- 2019-03-28 Makefile convenience variable 
- 2019-03-28 Minor configure changes. 
- 2019-03-28 Change default stack frame unwinding count. 
- 2019-03-18 Remove functions deprecated since MPI-2.0 
- 2019-03-18 Merge pull request #5 from bkmgit/master 
- 2019-02-20 Avoid LD_PRELOAD symbol conflict between srun and libiberty. 
- 2018-11-22 Create README.md 
- 2016-12-02 Corrected type issues resulting in compiler warnings in the test suite. 
- 2016-12-02 Make PC modification conditional. (Removed for now.) 
- 2016-12-02 Update config scripts. 
- 2016-05-19 Merge branch 'master' of https://github.com/LLNL/mpiP 
- 2016-05-19 Adding additional RMA functions to cover all MPI3.1 RMA functions.  Expanded test coverage. 
- 2016-05-19 Re-create configure to remove references to sourceforge. 
- 2016-05-19 Re-create configure to remove references to sourceforge. 
- 2016-05-16 Added MPI3 One-sided operation support, from changes by Jeff Hammond. 
- 2016-04-22 Merge branch 'master' of https://github.com/LLNL/mpiP 
- 2016-04-20 Final pass removing sourceforge references 
- 2016-04-20 Additional github migration. 
- 2016-04-19 Adding README.md 
- 2016-02-12 Added support for non-blocking collectives. 
- 2015-03-27 Minor cleanup. 
- 2015-03-27 Disable parallel makes. 
- 2015-03-20 Provide non-NULL error argument for dwarf calls. Remove call to elf_end.  Ignore compile units with no function information, rather than abort. 
- 2015-03-20 Reduce unnecessary debugging output. 
- 2015-03-20 Add count column to top time section. 
- 2015-03-20 Adding fortran symbol failure output for debugging configure issues. 
- 2014-03-13 mpiP-3.4.1 release tag. 
- 2014-03-13 mpiP-3.4.1 release commits. 
- 2013-08-01 mpiP-3.4 release. 
- 2013-08-01 Update ChangeLog. 
- 2013-08-01 Another pass with indent. 
- 2013-08-01 Minor testing changes. 
- 2013-08-01 Updating documentation to reflect new features. 
- 2013-07-18 Several small changes including: 
- 2013-07-17 Corrected errors related to MPI op-only reporting and COV calculation. 
- 2013-07-17 Added LDFLAGS to shared library build. 

util.c
/* -*- C -*- 

   mpiP MPI Profiler ( http://llnl.github.io/mpiP )

   Please see COPYRIGHT AND LICENSE information at the end of this file.

   -----

   util.c -- misc util functions

 */

#ifndef lint
static char *svnid = "$Id$";
#endif

#include <string.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>

#include "mpiPi.h"
#ifndef ENABLE_API_ONLY
#include "symbols.h"
#endif


static int argc = 0;
static char **argv = NULL;


char *
GetBaseAppName (char *rawName)
{
  char *cp;

  if (rawName == NULL)
    return strdup ("Unknown");

  /* delete directorys and make a copy, if nec */
  if ((cp = rindex (rawName, '/')) == NULL)
    {
      return rawName;
    }
  else
    {
      return cp + 1;		/* skip over the '/' */
    }
}


void
mpiPi_getenv ()
{  /* NEED TO ADD SANITY CHECKS HERE */
  char *cp = NULL;
  char *ep = NULL;

  mpiPi.outputDir = ".";

  ep = getenv ("MPIP");
  mpiPi.envStr = (ep ? strdup (ep) : 0);
  optind = 1;			/* reset to avoid conflicts if getopt already called */
  if (ep != NULL)
    {
      int c;
      extern char *optarg;
      int ac = 0;
      char *av[64];
      char *sep = " \t,";

      if (mpiPi.rank == 0)
        mpiPi_msg ("Found MPIP environment variable [%s]\n", ep);
      av[0] = "JUNK";
      for (cp = strtok (ep, sep), ac = 1; (ac < 64) && (NULL != cp); ac++)
        {
          av[ac] = cp;
          /*mpiPi_msg("av[%d] = %s\n", ac, av[ac]); */
          cp = strtok (NULL, sep);
        }

      av[ac] = NULL;

      for (; ((c = getopt (ac, av, "cdef:gk:lm:noprs:t:vx:yz")) != EOF);)
        {
          switch (c)
            {
            case 'f':
              mpiPi.outputDir = optarg;
              if (mpiPi.rank == 0)
                mpiPi_msg ("Set the output directory to [%s].\n",
                           mpiPi.outputDir);
              break;

            case 'g':
              mpiPi_debug = 1;
              if (mpiPi.rank == 0)
                mpiPi_msg ("Enabled mpiPi debug mode.\n");
              break;

            case 's':
              {
                int defaultSize = mpiPi.tableSize;
                mpiPi.tableSize = atoi (optarg);
                if (mpiPi.tableSize < 2)
                  {
                    if (mpiPi.rank == 0)
                      mpiPi_msg_warn
                          ("-s tablesize invalid %d. Using default.\n",
                           mpiPi.tableSize);
                    mpiPi.tableSize = defaultSize;
                  }
                if (mpiPi.tableSize < 128)
                  {
                    if (mpiPi.rank == 0)
                      mpiPi_msg_warn
                          ("tablesize small %d. Consider making it larger w/ -s.\n",
                           mpiPi.tableSize);
                  }
                if (mpiPi.rank == 0)
                  mpiPi_msg
                      ("Set the callsite table size to [%d].\n",
                       mpiPi.tableSize);
              }
              break;

            case 'k':
              {
                mpiPi.reportStackDepth = atoi (optarg);
                if (mpiPi.reportStackDepth < 0)
                  {
                    if (mpiPi.rank == 0)
                      mpiPi_msg_warn
                          ("-k stackdepth invalid %d. Using 0.\n",
                           mpiPi.reportStackDepth);
                    mpiPi.reportStackDepth = 0;
                    mpiPi.print_callsite_detail = 0;
                  }
                if (mpiPi.reportStackDepth > MPIP_CALLSITE_REPORT_STACK_DEPTH_MAX)
                  {
                    if (mpiPi.rank == 0)
                      mpiPi_msg_warn
                          ("stackdepth of %d too large. Using %d.\n",
                           mpiPi.reportStackDepth, MPIP_CALLSITE_REPORT_STACK_DEPTH_MAX);
                    mpiPi.reportStackDepth = MPIP_CALLSITE_REPORT_STACK_DEPTH_MAX;
                  }
                else if (mpiPi.reportStackDepth > 4)
                  {
                    if (mpiPi.rank == 0)
                      mpiPi_msg_warn
                          ("stackdepth of %d is large. Consider making it smaller.\n",
                           mpiPi.reportStackDepth);
                  }

                //  If the stack depth is 0, we are accumulating data
                //  for each MPI op (i.e. potentially multiple callsites),
                //  resulting in data that would not be useful for calculating COV.
                if (mpiPi.reportStackDepth == 0)
                  mpiPi.calcCOV = 0;

                if (mpiPi.rank == 0)
                  mpiPi_msg
                      ("Set the callsite stack traceback depth to [%d].\n",
                       mpiPi.reportStackDepth);
              }
              mpiPi.fullStackDepth = mpiPi.reportStackDepth + mpiPi.internalStackDepth;

              break;

            case 't':
              {
                int defaultThreshold = mpiPi.reportPrintThreshold;
                mpiPi.reportPrintThreshold = atof (optarg);
                if (mpiPi.reportPrintThreshold < 0)
                  {
                    if (mpiPi.rank == 0)
                      mpiPi_msg_warn
                          ("-t report print threshold invalid %g. Using default.\n",
                           mpiPi.reportPrintThreshold);
                    mpiPi.reportPrintThreshold = defaultThreshold;
                  }
                if (mpiPi.reportPrintThreshold >= 100)
                  {
                    if (mpiPi.rank == 0)
                      mpiPi_msg_warn
                          ("report print threshold large %g. Making it default.\n",
                           mpiPi.reportPrintThreshold);
                    mpiPi.reportPrintThreshold = defaultThreshold;
                  }
                if (mpiPi.rank == 0)
                  mpiPi_msg
                      ("Set the report print threshold to [%3.2lf%%].\n",
                       mpiPi.reportPrintThreshold);
              }
              break;

            case 'o':
              {
                if (mpiPi.rank == 0)
                  mpiPi_msg_warn
                      ("Disabling MPIP at Init. Code must use Pcontrol to enable.\n");
                mpiPi.enabled = 0;
                mpiPi.enabledCount = 0;
              }
              break;

            case 'n':
              mpiPi.baseNames = 1;
              break;

            case 'e':
              mpiPi.reportFormat = MPIP_REPORT_FLT_FORMAT;
              break;

            case 'c':
              mpiPi.report_style = mpiPi_style_concise;
              break;

            case 'v':
              mpiPi.report_style = mpiPi_style_both;
              break;

            case 'm':
              mpiPi.messageCountThreshold = atoi (optarg);
              mpiPi_msg_debug ("Set messageCountThreshold to %d\n",
                               mpiPi.messageCountThreshold);
              break;
            case 'x':
              if (optarg != NULL)
                {
                  mpiPi.appFullName = strdup (optarg);
                  mpiPi.av[0] = strdup (optarg);
                  mpiPi.appName = strdup (GetBaseAppName (mpiPi.appFullName));
                  mpiPi_msg_debug ("Set mpiPi.appFullName to %s\n",
                                   mpiPi.appFullName);
                }
              break;
            case 'd':
              /*  Suppress/Activate printing of call site detail based on default.  */
              mpiPi.print_callsite_detail ^= 1;
              break;

            case 'l':
              /*  Use low-memory use approach using MPI collectives
             for report generation   */
              mpiPi.collective_report = 1;
              break;

            case 'r':
              /*  Use collector task to aggregate all task data and
             generate report   */
              mpiPi.collective_report = 0;
              break;

            case 'z':
              mpiPi.disable_finalize_report = 1;
              break;

            case 'y':
              mpiPi.do_collective_stats_report = 1;
              break;

            case 'p':
              mpiPi.do_pt2pt_stats_report = 1;
              break;

            case 'a':
            case 'b':
            case 'h':
            case 'i':
            case 'j':
            case 'q':
            case 'u':
            case 'w':
            default:
              if (mpiPi.rank == 0)
                mpiPi_msg_warn
                    ("Option flag (-%c) not recognized. Ignored.\n", c);
              break;
            }
        }
    }
  if (mpiPi.rank == 0)
    mpiPi_msg ("\n");
  optind = 1;			/* reset to avoid conflicts if getopt called again */
}


#ifdef Linux
char *
getProcExeLink ()
{
  int pid, exelen, insize = 256;
  char *inbuf = NULL, file[256];

  pid = getpid ();
  snprintf (file, 256, "/proc/%d/exe", pid);
  inbuf = malloc (insize);
  if (inbuf == NULL)
    {
      mpiPi_abort ("unable to allocate space for full executable path.\n");
    }

  exelen = readlink (file, inbuf, 256);
  if (exelen == -1)
    {
      if (errno != ENOENT)
        {
          while (exelen == -1 && errno == ENAMETOOLONG)
            {
              insize += 256;
              inbuf = realloc (inbuf, insize);
              exelen = readlink (file, inbuf, insize);
            }
          inbuf[exelen] = '\0';
          return inbuf;
        }
      else
        free (inbuf);
    }
  else
    {
      inbuf[exelen] = '\0';
      return inbuf;
    }
  return NULL;
}

#define MPIP_MAX_ARG_STRING_SIZE 4096
void
getProcCmdLine (int *ac, char **av)
{
  int i = 0, pid;
  char *inbuf, file[256];
  FILE *infile;
  char *arg_ptr;

  *ac = 0;
  *av = NULL;

  pid = getpid ();
  snprintf (file, 256, "/proc/%d/cmdline", pid);
  infile = fopen (file, "r");

  if (infile != NULL)
    {
      while (!feof (infile))
        {
          inbuf = malloc (MPIP_MAX_ARG_STRING_SIZE);
          if (fread (inbuf, 1, MPIP_MAX_ARG_STRING_SIZE, infile) > 0)
            {
              arg_ptr = inbuf;
              while (*arg_ptr != '\0')
                {
                  av[i] = strdup (arg_ptr);
                  arg_ptr += strlen (av[i]) + 1;
                  i++;
                }
            }
        }
      *ac = i;

      free (inbuf);
      fclose (infile);
    }
}

#endif

void
mpiPi_copy_args (int *ac, char **av, int av_len)
{
  assert (ac != NULL);
  assert (av != NULL);

#if !defined(USE_GETARG)
#if defined(AIX)
  {
    /*  Works for C/C++ and Fortran  */
    extern int p_xargc;
    extern char **p_xargv;
    argc = p_xargc;
    argv = p_xargv;
  }
#elif defined(GNU_Fortran)
  {
    extern int f__xargc;
    extern char **f__xargv;
    argc = f__xargc;
    argv = f__xargv;
  }
#elif defined(OSF1)
  {
    /*  Works for C/C++ and Fortran  */
    extern int __Argc;
    extern char **__Argv;
    argc = __Argc;
    argv = __Argv;
  }
#elif defined(PGI)
  {
    extern int __argc_save;
    extern char **__argv_save;
    argc = __argc_save;
    argv = __argv_save;
  }
#endif
#endif /* USE_GETARG */

  mpiPi_copy_given_args (ac, av, av_len, argc, argv);
}


void
mpiPi_copy_given_args (int *ac, char **av, int av_len, int argc, char **argv)
{
  int i;

  assert (ac != NULL);
  assert (av != NULL);

#if defined(USE_GETARG)

#define EXECUTABLE_LEN 2048
  {
    char buf[EXECUTABLE_LEN];
    int len, mpiPi_argc;
    extern void F77_MPIPI_GET_FORTRAN_ARGC (int *);

    F77_MPIPI_GET_FORTRAN_ARGC (&mpiPi_argc);
    mpiPi_argc++;
    *ac = mpiPi_argc;

    for (i = 0; i < mpiPi_argc; i++)
      {
        int buf_len = EXECUTABLE_LEN;

        extern void F77_MPIPI_GET_FORTRAN_ARG (int *, int *, char *, int *,
                                               int);
        F77_MPIPI_GET_FORTRAN_ARG (&i, &buf_len, buf, &len, EXECUTABLE_LEN);

        buf[len < EXECUTABLE_LEN ? len : EXECUTABLE_LEN - 1] = 0;
        av[i] = strdup (buf);
      }
  }

#else /* USE_GETARG */
  *ac = argc;

  for (i = 0; (i < *ac) && (i < av_len); i++)
    av[i] = strdup (argv[i]);
#endif
}


char *
mpiP_format_address (void *pval, char *addr_buf)
{
  static int get_sys_info = 0;
  static int ptr_hex_chars = 0;
  static char hex_prefix[3] = "";
  char test_buf[8] = "";

  if (get_sys_info == 0)
    {
      ptr_hex_chars = sizeof (char *) * 2;
#ifdef Linux
      snprintf (test_buf, 8, "%p", (void *) 0x1);
#else
      snprintf (test_buf, 8, "%0p", (void *) 0x1);
#endif

      if (strcmp (test_buf, "0x1") != 0)
        strcpy (hex_prefix, "0x");

      get_sys_info = 1;
    }

#ifdef Linux
  sprintf (addr_buf, "%s%p", hex_prefix, pval);
#else
  sprintf (addr_buf, "%s%0.*p", hex_prefix, ptr_hex_chars, pval);
#endif

  return addr_buf;
}



/* 

<license>

Copyright (c) 2006, The Regents of the University of California. 
Produced at the Lawrence Livermore National Laboratory 
Written by Jeffery Vetter and Christopher Chambreau. 
UCRL-CODE-223450. 
All rights reserved. 

This file is part of mpiP.  For details, see http://llnl.github.io/mpiP. 

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

* Redistributions of source code must retain the above copyright
notice, this list of conditions and the disclaimer below.

* Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the disclaimer (as noted below) in
the documentation and/or other materials provided with the
distribution.

* Neither the name of the UC/LLNL nor the names of its contributors
may be used to endorse or promote products derived from this software
without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OF
THE UNIVERSITY OF CALIFORNIA, THE U.S. DEPARTMENT OF ENERGY OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


Additional BSD Notice 

1. This notice is required to be provided under our contract with the
U.S. Department of Energy (DOE).  This work was produced at the
University of California, Lawrence Livermore National Laboratory under
Contract No. W-7405-ENG-48 with the DOE.

2. Neither the United States Government nor the University of
California nor any of their employees, makes any warranty, express or
implied, or assumes any liability or responsibility for the accuracy,
completeness, or usefulness of any information, apparatus, product, or
process disclosed, or represents that its use would not infringe
privately-owned rights.

3.  Also, reference herein to any specific commercial products,
process, or services by trade name, trademark, manufacturer or
otherwise does not necessarily constitute or imply its endorsement,
recommendation, or favoring by the United States Government or the
University of California.  The views and opinions of authors expressed
herein do not necessarily state or reflect those of the United States
Government or the University of California, and shall not be used for
advertising or product endorsement purposes.

</license>

*/


/* eof */
